### PR TITLE
Simple POST

### PR DESCRIPTION
**Pull request by kelseyroy**
**Scenario:** Posting echos the body
* _Given_ I make a POST with a body to `/echo_body`
* _Then_ my response should have status code 200
* _And_ my response body should equal the body of my request

Adds:
* Request & Response as structs
* `parse` to handle parsing the HTTP request into Request struct
* `format_headers` to convert request headers into a map
* `build_response` to build a Response struct with a request struct
* `text` to format Response struct as strings
* `response_headers` to format response headers into strings

Changes:
* `read_line` to ensure process can end without throwing errors if client closes
* `serve` to incorporate these new processes

Potential Issues/what I may need help with:
* There is currently no functionality that is decoding or encoding, yet the acceptance tests are passing and I can send a request with Postman. I, truthfully, don't understand why it's working without encode/decode.
* Is there an opportunity for better naming? Some of the functions, like `format_headers` and `response_headers` feel like they serve similar purposes.

### CODE REVIEW

**Comment by kelseyroy** [comment]
[Simple POST](https://trello.com/c/pXYYYbNc/18-simple-post)

File: lib/http_server.ex
Comment (Aoxian) [inline lib/http_server.ex]: maybe `format_response`?

File: lib/http_server.ex
Comment (Aoxian) [inline lib/http_server.ex]: maybe `parse_headers`?

File: test/http_server_test.exs
Comment (Aoxian) [inline test/http_server_test.exs:12]: why not write this test now?

File: test/http_server_test.exs
Comment (Aoxian) [inline test/http_server_test.exs]: type not needed.
```suggestion
      request = %HTTPServer.Request{
```

```suggestion
      test_request = %HTTPServer.Request{
```

File: test/http_server_test.exs
Comment (Aoxian) [inline test/http_server_test.exs]:
```suggestion
      expected_parsed_request = %HTTPServer.Request{
```

File: test/http_server_test.exs
Comment (Aoxian) [inline test/http_server_test.exs]:
```suggestion
      expected_response = %HTTPServer.Response{
```

Comment (Aoxian) [reply test/http_server_test.exs:59]: same in other tests

File: test/http_server_test.exs
Comment (Aoxian) [inline test/http_server_test.exs]:
```suggestion
      expected_response_as_text =
```

File: test/http_server_test.exs
Comment (Aoxian) [inline test/http_server_test.exs]: type not needed

```suggestion
      response = %HTTPServer.Response{
```

```suggestion
      test_response_struct = %HTTPServer.Response{
```

**Comment by Aoxian** [comment]
Looking super "Elixiry". Love it!

File: lib/http_server.ex
Comment (t-keazirian) [inline lib/http_server.ex]: I see a lot of hard-coded "\r\n" around here - maybe time to extract it into a variable?

File: test/http_server_test.exs
Comment (t-keazirian) [inline test/http_server_test.exs]: 👀  You have some GitHub Actions alerts/failures here.

Comment (kelseyroy) [reply test/http_server_test.exs]: Ooooh thanks for calling this out! I was committing some of the name suggestions but didn't change all instances. I think it should be fixed now!

File: lib/http_server.ex
Comment (t-keazirian) [reply lib/http_server.ex]: I like how most of this is extracted - I wonder if there's a way you could tie the status_code to the word that follows? (e.g. 200 => OK, 201 => Created, etc.)

File: lib/http_server.ex
Comment (t-keazirian) [inline lib/http_server.ex]: This seems like it might be formatting headers into key-value pairs. Would `format_response_headers` be a more explicit name? Just food for thought if it makes sense in this case? 

Then it would read like: `format_response_headers(res.headers)`

File: lib/http_server.ex
Comment (t-keazirian) [inline lib/http_server.ex]:
```suggestion
  def parse_request(message) do
```
if I'm reading it correctly, this is parsing the request - is that right? if so, this might help with clarity so I know at first glance.

File: lib/http_server.ex
Comment (t-keazirian) [inline lib/http_server.ex:1]: This module is getting a little long - wondering if everything in here belongs in here or if any of the functions could be extracted into their own module(s)? Thoughts?

File: lib/http_server.ex
Comment (Aoxian) [reply lib/http_server.ex]: Maybe for a future PR, but I think there maybe some modules hiding out here. All the domain language is begging for it.

File: lib/http_server.ex
Comment (Aoxian) [reply lib/http_server.ex:1]: Same [thought](https://github.com/kelseyroy/http_server/pull/3#discussion_r1213342021), at the same time

Comment (Aoxian) [reply lib/http_server.ex:1]: image

File: lib/http_server.ex
Comment (t-keazirian) [reply lib/http_server.ex]: also, if it's parsing the request, maybe parameter named `request` is more explicit?

File: lib/http_server.ex
Comment (kelseyroy) [reply lib/http_server.ex]: @t-keazirian it's taking the key value pairs from the headers map and putting them into strings... so I agree that `format_response_headers` is a better name to describe the functionality!

File: lib/http_server.ex
Comment (kelseyroy) [reply lib/http_server.ex:1]: Totally agree about needing to break down this module. I was thinking of adding some of these formatting/parsing functions to the existing request/response (depending on which one they deal with).

File: lib/http_server.ex
Comment (t-keazirian) [reply lib/http_server.ex]: the way I like to think about it is - functions --> verbs, variables --> nouns (generally speaking). so this stood out to me as a code smell because `response_headers` is a noun, but this is a function performing an action, so the name and action didn't align 😄

File: test/http_server_test.exs
Comment (t-keazirian) [inline test/http_server_test.exs:39]:
```suggestion
      actual_parsed_request = HTTPServer.parse_request(message)
      assert actual_parsed_request == expected_parsed_request
```

thoughts about pulling the calls out into a variable and comparing the two (here and in the other tests)?

**Comment by t-keazirian** [comment]
> Looking super "Elixiry". Love it!

Agree!!!  So fun to see!

File: lib/http_server.ex
Comment (t-keazirian) [inline lib/http_server.ex:55]: 👀 you might find you'll want a `helpers` or `constants` module for `carriage_return` (and potentially `" "`) since you're using it in multiple places/functions (I see you already extracted it below - nice!)

**Review by Aoxian — APPROVED** [review]
Lots of great review comments here; I do not think they ALL need to be incorporated into this one PR. I am approving so this can get merged and demo'd for IPM.

Lesson: Sometimes, you ship-it because of deadlines!

**Review by Aoxian — APPROVED** [review]
Lots of great review comments here; I do not think they ALL need to be incorporated into this one PR. I am approving so this can get merged and demo'd for IPM.

Lesson: Sometimes, you ship-it because of deadlines!

**Review by Aoxian — APPROVED** [review]
Lots of great review comments here; I do not think they ALL need to be incorporated into this one PR. I am approving so this can get merged and demo'd for IPM.

Lesson: Sometimes, you ship-it because of deadlines!

**Review by Aoxian — APPROVED** [review]
Lots of great review comments here; I do not think they ALL need to be incorporated into this one PR. I am approving so this can get merged and demo'd for IPM.

Lesson: Sometimes, you ship-it because of deadlines!

**Review by Aoxian — APPROVED** [review]
Lots of great review comments here; I do not think they ALL need to be incorporated into this one PR. I am approving so this can get merged and demo'd for IPM.

Lesson: Sometimes, you ship-it because of deadlines!

**Review by Aoxian — APPROVED** [review]
Lots of great review comments here; I do not think they ALL need to be incorporated into this one PR. I am approving so this can get merged and demo'd for IPM.

Lesson: Sometimes, you ship-it because of deadlines!

**Review by Aoxian — APPROVED** [review]
Lots of great review comments here; I do not think they ALL need to be incorporated into this one PR. I am approving so this can get merged and demo'd for IPM.

Lesson: Sometimes, you ship-it because of deadlines!

**Review by Aoxian — APPROVED** [review]
Lots of great review comments here; I do not think they ALL need to be incorporated into this one PR. I am approving so this can get merged and demo'd for IPM.

Lesson: Sometimes, you ship-it because of deadlines!

**Comment by Aoxian** [comment]
^ haha, github bug? you just got 8 approvals, lol

File: lib/http_server.ex
Comment (kelseyroy) [reply lib/http_server.ex:55]: Oooooh this is a helpful nudge in the right direction, because I changed it in `format_response` but realized I use it a lot outside of those functions as well.

**Pull request merged**
